scrapyd_k8s/k8s_resource_watcher.py

```diff
@@ -0,0 +1,106 @@
+import threading
+import logging
+import time
+from kubernetes import client, watch
+from typing import Callable, List
+
+logger = logging.getLogger(__name__)
+
+class ResourceWatcher:
+    """
+    Watches Kubernetes pod events and notifies subscribers about relevant events.
+
+    Attributes
+    ----------
+    namespace : str
+        Kubernetes namespace to watch pods in.
+    subscribers : List[Callable]
+        List of subscriber callback functions to notify on events.
+    """
+
+    def __init__(self, namespace):
+        """
+        Initializes the ResourceWatcher.
+
+        Parameters
+        ----------
+        namespace : str
+            Kubernetes namespace to watch pods in.
+        """
+        self.namespace = namespace
+        self.subscribers: List[Callable] = []
+        self._stop_event = threading.Event()
+        self.watcher_thread = threading.Thread(target=self.watch_pods, daemon=True)
+        self.watcher_thread.start()
+        logger.info(f"ResourceWatcher thread started for namespace '{self.namespace}'.")
+
+    def subscribe(self, callback: Callable):
+        """
+        Adds a subscriber callback to be notified on events.
+
+        Parameters
+        ----------
+        callback : Callable
+            A function to call when an event is received.
+        """
+        if callback not in self.subscribers:
+            self.subscribers.append(callback)
+            logger.debug(f"Subscriber {callback.__name__} added.")
+
+    def unsubscribe(self, callback: Callable):
+        """
+        Removes a subscriber callback.
+
+        Parameters
+        ----------
+        callback : Callable
+            The subscriber function to remove.
+        """
+        if callback in self.subscribers:
+            self.subscribers.remove(callback)
+            logger.debug(f"Subscriber {callback.__name__} removed.")
+
+    def notify_subscribers(self, event: dict):
+        """
+        Notifies all subscribers about an event.
+
+        Parameters
+        ----------
+        event : dict
+            The Kubernetes event data.
+        """
+        for subscriber in self.subscribers:
+            try:
+                subscriber(event)
+            except Exception as e:
+                logger.exception(f"Error notifying subscriber {subscriber.__name__}: {e}")
```

Comment:
why catch this? when do you expect this?

Reply:
This exception is added so we can separate the exception originated in particular subscriber from watcher and from other subscribers.
If there is an unexpected edge case for subscriber it is nice to catch it and understand where it comes from, also easier to debug.

Reply:
I think the broad `Exception` should be part of this try-except clause since it is unpredictable what callbacks are executed (not really since we only have 2 currently, but in terms of future proofness). Also, one of the callbacks also raises a bare `Exception` when an issue is encountered. We also need to make sure that when one of the subscribers fails, we do send the event to the other subscribers.

However, it might be better to also re-raise that bare `Exception` as a custom exception so we can do more specific error handling and logging. Same goes for the other callback. Maybe we can design a custom exception that can be raised in the callback that we can then catch while executing these callbacks.

Reply:
Thank you for your input, colleagues!
What I see here is:
1. Keep the broad exception for the Observer since we cannot guarantee what subscribers do and we do not want one subscriber to mess up all the features that rely on the Observer and Observer itself.
2. Think of custom exceptions for subscribers. - this is something I am going to work on.